packages/components/form/src/form.vue

```diff
@@ -1,258 +1,158 @@
-      })
-    }
+  for (const field of fds) {
+    const fieldsError = await field
+      .validate('')
+      .catch((fields: ValidateFieldsError) => fields)
 
-    const clearValidate = (props: string | string[] = []) => {
-      const fds = props.length
-        ? typeof props === 'string'
-          ? fields.filter((field) => props === field.prop)
-          : fields.filter((field) => props.indexOf(field.prop) > -1)
-        : fields
-      fds.forEach((field) => {
-        field.clearValidate()
-      })
+    if (fieldsError) {
+      valid = false
+      if (!firstInvalidFields) firstInvalidFields = fieldsError
     }
 
-    const validate = (callback?: Callback) => {
-      if (!props.model) {
-        debugWarn('Form', 'model is required for validate to work!')
-        return
-      }
-
-      let promise: Promise<boolean> | undefined
-      // if no callback, return promise
-      if (typeof callback !== 'function') {
-        promise = new Promise((resolve, reject) => {
-          callback = function (valid, invalidFields) {
-            if (valid) {
-              resolve(true)
-            } else {
-              reject(invalidFields)
-            }
-          }
-        })
-      }
-
-      if (fields.length === 0) {
-        callback(true)
-      }
-      let valid = true
-      let count = 0
-      let invalidFields = {}
-      let firstInvalidFields
-      for (const field of fields) {
-        field.validate('', (message, field) => {
-          if (message) {
-            valid = false
-            firstInvalidFields || (firstInvalidFields = field)
-          }
-          invalidFields = { ...invalidFields, ...field }
-          if (++count === fields.length) {
-            callback(valid, invalidFields)
-          }
-        })
-      }
-      if (!valid && props.scrollToError) {
-        scrollToField(Object.keys(firstInvalidFields)[0])
-      }
-      return promise
-    }
+    invalidFields = { ...invalidFields, ...fieldsError }
+  }
 
-    const validateField = (
-      props: string | string[],
-      cb: ValidateFieldCallback
-    ) => {
-      props = [].concat(props)
-      const fds = fields.filter((field) => props.indexOf(field.prop) !== -1)
-      if (!fields.length) {
-        debugWarn('Form', 'please pass correct props!')
-        return
-      }
+  if (!valid) {
+    if (scrollToError) scrollToField(Object.keys(firstInvalidFields!)[0])
+    return Promise.reject(invalidFields)
+  }
+}
 
-      fds.forEach((field) => {
-        field.validate('', cb)
-      })
-    }
+const scrollToField = (prop: string) => {
+  const field = filterFields(fields, prop)[0]
+  if (field) {
+    field.$el?.scrollIntoView()
+  }
+}
 
-    const scrollToField = (prop: string) => {
-      fields.forEach((item) => {
-        if (item.prop === prop) {
-          item.$el.scrollIntoView?.()
-        }
-      })
-    }
+watch(
+  () => props.rules,
+  () => {
+    if (props.validateOnRuleChange) validate()
+  }
+)
```

Comment:
**Suggestion:** for robust code here I think we should deep watch `props.rules` since `Vue` won't automatically deep watch for us.